diskqueue.go

```diff
@@ -395,6 +401,223 @@ func (d *diskQueue) readOne() ([]byte, error) {
+			// we want to see the contents of the directory DiskQueue writes and reads in
+			return nil
+		}
+
+		if dirEntry.IsDir() {
+			// if the entry is a directory, skip it
+			return fs.SkipDir
+		}
+
+		if err != nil {
+			return err
+		}
+
+		return fn(dirEntry)
+	}
+
+	err := filepath.WalkDir(d.dataPath, walkDiskQueueDir)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *diskQueue) getAllBadFileInfo() ([]fs.FileInfo, error) {
+	var badFileInfos []fs.FileInfo
+
+	getAllBadFileInfo := func(dirEntry fs.DirEntry) error {
+		// only accept "bad" files created by this DiskQueue object
+		regExp, _ := regexp.Compile(`^` + d.name + `.diskqueue.\d\d\d\d\d\d.dat.bad$`)
+
+		if regExp.MatchString(dirEntry.Name()) {
+			badFileInfo, e := dirEntry.Info()
+			if e == nil && badFileInfo != nil {
+				badFileInfos = append(badFileInfos, badFileInfo)
+			}
+		}
+
+		return nil
+	}
+
+	err := d.walkDiskQueueDir(getAllBadFileInfo)
+
+	return badFileInfos, err
+}
+
+// get the accurate total non-"bad" file size
+func (d *diskQueue) updateWriteBytes() error {
+	d.writeBytes = 0
+
+	updateWriteBytes := func(dirEntry fs.DirEntry) error {
+		// only accept files created by this DiskQueue object
+		regExp, _ := regexp.Compile(`^` + d.name + `.diskqueue.\d\d\d\d\d\d.dat$`)
+		if regExp.MatchString(dirEntry.Name()) {
+			fileInfo, e := dirEntry.Info()
+			if e == nil && fileInfo != nil {
+				d.writeBytes += fileInfo.Size()
+			}
+		}
+
+		return nil
+	}
+
+	return d.walkDiskQueueDir(updateWriteBytes)
+}
+
+// free up disk space if needed to write new data to file
+func (d *diskQueue) freeUpDiskSpace(totalBytes int64, reachedFileSizeLimit bool) error {
+	var err error
+
+	expectedBytesIncrease := totalBytes
+	if reachedFileSizeLimit {
+		expectedBytesIncrease += numFileMsgBytes
+	}
+
+	// If the data to be written is bigger than the disk size limit, do not write
+	if expectedBytesIncrease > d.maxBytesDiskSpace {
+		return errors.New("message size surpasses disk size limit")
+	}
+
+	// check if we have enough space to write this message
+	metaDataFileSize := d.metaDataFileSize()
+
+	// get total disk space of bad files
+	var badFilesSize int64
+	badFileInfos, err := d.getAllBadFileInfo()
+	if err != nil {
+		d.logf(ERROR, "DISKQUEUE(%s) failed to retrieve all .bad file info - %s", d.name, err)
+	}
+	for _, badFileInfo := range badFileInfos {
+		badFilesSize += badFileInfo.Size()
+	}
+
+	// keep freeing up disk space until we have enough space to write this message
+	for badFilesSize+metaDataFileSize+d.writeBytes+expectedBytesIncrease > d.maxBytesDiskSpace {
+		if badFilesSize > 0 {
+			// check if a .bad file exists. If it does, delete that first
+			oldestBadFileInfo := badFileInfos[0]
+			badFileFilePath := path.Join(d.dataPath, oldestBadFileInfo.Name())
+
+			err = os.Remove(badFileFilePath)
+			if err != nil {
+				d.logf(ERROR, "DISKQUEUE(%s) failed to remove .bad file(%s) - %s", d.name, oldestBadFileInfo.Name(), err)
+				return err
+			} else {
```

Comment:
Nice catch! I wasn't sure if we needed to account for external agents possibly deleting or renaming ".bad" files, but I agree that it would be more efficient to decrease `badFileSize` by `oldestBadFileInfo.Size()` and set `badFileInfos` to `badFileInfos[1:]`